bundle/src/main/java/com/cognifide/cq/cqsm/core/utils/sling/SlingHelper.java

```diff
@@ -125,22 +135,53 @@ public static void operateTraced(ResourceResolverFactory factory, String userId,
 		}
 	}
 
-	/**
-	 * Create a new session for specified user (impersonating).
-	 */
-	public static ResourceResolver getResourceResolverForUser(ResourceResolverFactory factory, String userId)
-			throws LoginException {
-		ResourceResolver resolver;
-		if (userId != null) {
-			Map<String, Object> authenticationInfo = Maps.newHashMap();
-			authenticationInfo.put(ResourceResolverFactory.USER_IMPERSONATION, userId);
-			resolver = factory.getAdministrativeResourceResolver(authenticationInfo);
-		} else {
-			Map<String, Object> parameters = new HashMap<>();
-			parameters.put(ResourceResolverFactory.SUBSERVICE, SUBSERVICE_NAME);
-			resolver = factory.getServiceResourceResolver(parameters);
-		}
-
-		return resolver;
-	}
+    /**
+     * Create a new session for specified user (impersonating).
+     */
+    public static ResourceResolver getResourceResolverForUser(ResourceResolverFactory factory, String userId)
+            throws LoginException, RepositoryException {
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put(ResourceResolverFactory.SUBSERVICE, SUBSERVICE_NAME);
+        ResourceResolver resolver = factory.getServiceResourceResolver(parameters);
+        if (userId != null) {
+            Session session = resolver.adaptTo(Session.class);
+            addCqsmUserAsImpersonator(session, userId);
+            resolver = impersonate(factory, userId, session);
+        }
+
+        return resolver;
+    }
+
+    private static void addCqsmUserAsImpersonator(Session session, String userId) throws RepositoryException {

```

Comment:
for `addXXX` methods usually we are expecting that adding will be done always, not conditionally. Personally I prefer for such methods `ensure` as prefix